# Math project 1

In [3]:
import numpy as np
import pandas as pd

In [4]:
df= pd.read_csv("insurance .csv")
df

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

# 1.Explain the data. Check It is supervised or Unsupervised. Demonstrate?

ANS:
We are having age,bmi,sex,children,smoker,region, charges column

Age,bmi – numerical data
Sex,children,smoker,region are categorical data

It is an supervised data. As it is labelled data 

# 2. Estimate the minimum sample size n to get the 99% accurate predictions. (precision = 0.02)

sample size is n = (z**2/e**2)*p*q
p=.5,z at 99% is -2.575, e=0.02

In [5]:


z=-2.57
e=0.02
p=0.5
q=1-p

n= (z**2/e**2)*p*q
print("Sample size is :",n)

Sample size is : 4128.062499999999


# 3. Check the data is cleaned or not. If not then clean it (Null values, Row/Column Duplicates,Outliers, Change the string into numbers)


In [6]:
# NULL values
df.isnull().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

In [7]:
# Row/Column Duplicates
duplicate= df[df.duplicated()]

In [8]:
duplicate


age   sex    bmi  children smoker     region    charges
581   19  male  30.59         0     no  northwest  1639.5631

In [9]:
df.drop_duplicates(inplace=True)

In [10]:
df.duplicated().sum()

0

In [11]:
# Outliers
df.describe().T

count          mean           std        min       25%        50%  \
age       1337.0     39.222139     14.044333    18.0000    27.000    39.0000   
bmi       1337.0     30.663452      6.100468    15.9600    26.290    30.4000   
children  1337.0      1.095737      1.205571     0.0000     0.000     1.0000   
charges   1337.0  13279.121487  12110.359656  1121.8739  4746.344  9386.1613   

                  75%          max  
age          51.00000     64.00000  
bmi          34.70000     53.13000  
children      2.00000      5.00000  
charges   16657.71745  63770.42801

In [12]:
df.dtypes

age           int64
sex          object
bmi         float64
children      int64
smoker       object
region       object
charges     float64
dtype: object

In [13]:
from scipy import stats
Q1 = df.quantile(0.25)
Q3 = df.quantile(0.75)
IQR = Q3 - Q1

print(IQR)

age            24.00000
bmi             8.41000
children        2.00000
charges     11911.37345
dtype: float64


In [14]:
df.notnull().sum()

age         1337
sex         1337
bmi         1337
children    1337
smoker      1337
region      1337
charges     1337
dtype: int64

In [15]:
df_outlier=df[(df < (Q1 - 1.5 * IQR)) |(df > (Q3 + 1.5 * IQR))]

<ipython-input-15-4dd03b02a3df>:1: FutureWarning: Automatic reindexing on DataFrame vs Series comparisons is deprecated and will raise ValueError in a future version.  Do `left, right = left.align(right, axis=1, copy=False)` before e.g. `left == right`
  df_outlier=df[(df < (Q1 - 1.5 * IQR)) |(df > (Q3 + 1.5 * IQR))]
<ipython-input-15-4dd03b02a3df>:1: FutureWarning: Automatic reindexing on DataFrame vs Series comparisons is deprecated and will raise ValueError in a future version.  Do `left, right = left.align(right, axis=1, copy=False)` before e.g. `left == right`
  df_outlier=df[(df < (Q1 - 1.5 * IQR)) |(df > (Q3 + 1.5 * IQR))]


In [16]:
df_outlier.isnull().sum()

age         1337
sex         1337
bmi         1328
children    1337
smoker      1337
region      1337
charges     1198
dtype: int64

In [17]:
df_o = df[~((df< (Q1 - 1.5 * IQR)) |(df> (Q3 + 1.5 * IQR))).any(axis=1)]
df_o.shape

<ipython-input-17-c1fa06add1cd>:1: FutureWarning: Automatic reindexing on DataFrame vs Series comparisons is deprecated and will raise ValueError in a future version.  Do `left, right = left.align(right, axis=1, copy=False)` before e.g. `left == right`
  df_o = df[~((df< (Q1 - 1.5 * IQR)) |(df> (Q3 + 1.5 * IQR))).any(axis=1)]
<ipython-input-17-c1fa06add1cd>:1: FutureWarning: Automatic reindexing on DataFrame vs Series comparisons is deprecated and will raise ValueError in a future version.  Do `left, right = left.align(right, axis=1, copy=False)` before e.g. `left == right`
  df_o = df[~((df< (Q1 - 1.5 * IQR)) |(df> (Q3 + 1.5 * IQR))).any(axis=1)]


(1192, 7)

In [18]:
df.shape

(1337, 7)

In [19]:
df_old=df
df=df_o
df

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1192 rows x 7 columns]

In [20]:
df.shape

(1192, 7)

In [21]:
# Change the string into numbers

In [22]:
df['sex']=df['sex'].map({'male':1,'female':0}).astype(int)

<ipython-input-22-06b48e35b479>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['sex']=df['sex'].map({'male':1,'female':0}).astype(int)


In [23]:
df

age  sex     bmi  children smoker     region      charges
0      19    0  27.900         0    yes  southwest  16884.92400
1      18    1  33.770         1     no  southeast   1725.55230
2      28    1  33.000         3     no  southeast   4449.46200
3      33    1  22.705         0     no  northwest  21984.47061
4      32    1  28.880         0     no  northwest   3866.85520
...   ...  ...     ...       ...    ...        ...          ...
1333   50    1  30.970         3     no  northwest  10600.54830
1334   18    0  31.920         0     no  northeast   2205.98080
1335   18    0  36.850         0     no  southeast   1629.83350
1336   21    0  25.800         0     no  southwest   2007.94500
1337   61    0  29.070         0    yes  northwest  29141.36030

[1192 rows x 7 columns]

In [24]:
df['smoker']=df['smoker'].map({'no':0,'yes':1}).astype(int)

<ipython-input-24-f382b2ad8c5e>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['smoker']=df['smoker'].map({'no':0,'yes':1}).astype(int)


In [25]:
df

age  sex     bmi  children  smoker     region      charges
0      19    0  27.900         0       1  southwest  16884.92400
1      18    1  33.770         1       0  southeast   1725.55230
2      28    1  33.000         3       0  southeast   4449.46200
3      33    1  22.705         0       0  northwest  21984.47061
4      32    1  28.880         0       0  northwest   3866.85520
...   ...  ...     ...       ...     ...        ...          ...
1333   50    1  30.970         3       0  northwest  10600.54830
1334   18    0  31.920         0       0  northeast   2205.98080
1335   18    0  36.850         0       0  southeast   1629.83350
1336   21    0  25.800         0       0  southwest   2007.94500
1337   61    0  29.070         0       1  northwest  29141.36030

[1192 rows x 7 columns]

In [26]:
df.region.unique()

array(['southwest', 'southeast', 'northwest', 'northeast'], dtype=object)

In [27]:
df['region']=df['region'].map({'southwest':1,'southeast':2,'northwest':3,'northeast':4}).astype(int)

<ipython-input-27-6945f87f88de>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['region']=df['region'].map({'southwest':1,'southeast':2,'northwest':3,'northeast':4}).astype(int)


In [28]:
df

age  sex     bmi  children  smoker  region      charges
0      19    0  27.900         0       1       1  16884.92400
1      18    1  33.770         1       0       2   1725.55230
2      28    1  33.000         3       0       2   4449.46200
3      33    1  22.705         0       0       3  21984.47061
4      32    1  28.880         0       0       3   3866.85520
...   ...  ...     ...       ...     ...     ...          ...
1333   50    1  30.970         3       0       3  10600.54830
1334   18    0  31.920         0       0       4   2205.98080
1335   18    0  36.850         0       0       2   1629.83350
1336   21    0  25.800         0       0       1   2007.94500
1337   61    0  29.070         0       1       3  29141.36030

[1192 rows x 7 columns]

# 4. Check that sex and smoking are statistically independent or not.

In [29]:
df.corr()

age       sex       bmi  children    smoker    region   charges
age       1.000000 -0.019950  0.122817  0.038948 -0.067880 -0.004032  0.436059
sex      -0.019950  1.000000  0.016352  0.016284  0.014355  0.006179 -0.021548
bmi       0.122817  0.016352  1.000000  0.006874 -0.256608 -0.152963 -0.060969
children  0.038948  0.016284  0.006874  1.000000 -0.006307 -0.022434  0.082308
smoker   -0.067880  0.014355 -0.256608 -0.006307  1.000000  0.047671  0.602377
region   -0.004032  0.006179 -0.152963 -0.022434  0.047671  1.000000  0.090858
charges   0.436059 -0.021548 -0.060969  0.082308  0.602377  0.090858  1.000000

In [30]:
df.age.corr(df.smoker)

-0.0678798200218371

from above we can say that sex and smoking are statistically independent

In [35]:
from scipy.stats import chi2_contingency
crosstable=pd.crosstab(df['sex'],df['smoker'])
crosstable
 #in our code 0 represent male and 1 represent female 
 #smoker yes =1 and no =0
o_val=crosstable.values
print(o_val)
chisq,pval,degfre,e_val=chi2_contingency(crosstable)
print(chisq)
print(pval)
print(degfre)
print(e_val)
chi_stat=(((o_val-e_val)**2)/e_val).sum().sum()


[[543  68]
 [511  70]]
0.16408145769952764
0.6854262817955425
1
[[540.26342282  70.73657718]
 [513.73657718  67.26342282]]


# from above we are seeing that pvalue is 0.6854 which is > than 0.05 so we cannot reject null hypo saying that sex and smoking are dependent 

In [ ]:
'''crosstable=pd.crosstab(df['sex'],df['smoker'])
crosstable
 #in our code 0 represent male and 1 represent female 
 #smoker yes =1 and no =0
o_val=crosstable.values
o_val
chisq,pval,degfre,e_val=chi2_contingency(crosstable)
chisq,pval,degfre
e_val
chi_stat=(((o_val-e_val)**2)/e_val).sum().sum()'''

# 5. Check that all regressor variables (independent variable) are independent of each other or not.



In [ ]:
df.corr()

from above table we can say that age,sex,bmi,children,smoker,region all column are independeent of each other

# 6. Check the dependency between response and regressors.


In [ ]:
df.corr()

from above table we can say that, 
smoker and age colum positive correlation on response
sex and bmi are weak negatively 
children and region are weak positive


# 7. Predict the regression Line to predict the charges for insurance using independent variables.


In [ ]:

from sklearn.linear_model import LinearRegression as LR

x=df[['age','sex','bmi','smoker']]
y=df.charges
model=LR()
model.fit(x,y)
model.coef_

In [ ]:
model.intercept_

pred=-2869.6239330281915+(244.33588129*age)+(-321.25280466*sex)+(58.34184004*bmi)+( 14654.80151362*smoker)

# 8. Predict the accuracy of the regression Model.


In [ ]:
from sklearn.model_selection import train_test_split

x=x=df[['age','sex','bmi','smoker']].values
y=df.charges.values
print("x is :",x)
print("y is : ",y)
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.25,random_state=0)
l= LR()
l.fit(x_train,y_train)
predict=l.predict(x_test)
accuracy=l.score(x_test,y_test)

print("Accuracy is",accuracy*100)

# 9. Predict insurance charge for Age = 29, Sex = F, bmi = 28, children = 1, Smoke = Yes, region = southeast.



In [ ]:

x=df[['age','sex','bmi','smoker','region','children']]
y=df.charges
model=LR()
model.fit(x,y)
print(model.coef_)
age = 29
sex = 1 # in our code 0 represent male and 1 represent female 
bmi = 28
smoker = 1 # smoker yes =1 and no =0
region = 2  #'southwest':1,'southeast':2,'northwest':3,'northeast':4
children=1
pred=-2869.6239330281915+(244.33588129*age)+(-321.25280466*sex)+(58.34184004*bmi)+( 14654.80151362*smoker)+( 473.48673395*region)+(415.47548188*children)
pred

# 10. Give the percentage of error in regression model.


In [ ]:
y= df[['charges']].values
d=0
y_mean = df.charges.mean() 
n =df.shape[0]
for i in df[['charges']].values:
    d = d + (i - y_mean)

mse = (d*d)/n  

print("MSE: " , mse)

In [ ]:
d.shape

# 11. Give the 95% confidence interval for average charge insurance.
 
x_mean +/-t*s /sqrt(n) with t value @ alpha=0.05 and n-1 degree freedom (1192-1) is 1.961957807


In [ ]:
t = 1.961957807
x_mean =df.charges.mean()
s = df.charges.std()
n =df.shape[0]
lower =x_mean -((t*s)/(n**0.5))
print("lower",lower)
upper =x_mean +((t*s)/(n**0.5))
print("upper",upper)

we are 95% confident that true pouplation mean lies between 9537.220930502175 and 10361.249148439103